### DISCUSSION TITLE
Pore-size distribution of CT images

### DISCUSSION THREAD

**Discussion by fsamaraie** [PMEAL/OpenPNM#2726] [Q&A] [answered]
Hello community!
I have been interested in calculating pores properties from some XCT image stacks and comparing results from different algorithms and software. I recently started using OpenPNM, so this question might sound simple. 
If I want to find some statistics (min,  max, average) of pore equivalent diameters, does this code do the job?

im=imageio.volread('MyData.tif')
snow_output = ps.networks.snow2(im, voxel_size=1, boundary_width=0)
pn = op.io.network_from_porespy(snow_output.network)
a=pn['pore.equivalent_diameter']
np.max(a), np.min(a), np.average(a)

I am asking this as I noticed that usually following these lines gives quite similar results to other pores extraction software, only there seems to be slightly more small pores from OpenPNM. I understand that different algorithms may give different results, I just want to make sure that this is the reason.

Thank you

**Comment by jgostick** [comment]
Network extraction algorithms are all different and sensitive to different features of the image, so I don't think that looking at the psd of a network is the right approach.  You're better off just analysing the image directly, with local thickness or chord length distributions, for example.  The [porespy docs](https://porespy.org/_examples/filters/index.html) has a few examples of things you can do.

**Comment by fsamaraie** [comment]
Thank you very much sir @jgostick. I am interested in finding the PSD as I am trying to relate it to other properties such as reactivity or pore clogging. That is why I wanted to find the PSD and Im not sure if just using the command a=pn['pore.equivalent_diameter'] suffices.
Appreciated,

**Reply by jgostick** [reply, marked as answer]
The PSD of the network is best found using something simple like `plt.hist(pn['pore.equivalent_diameter'], density=True)`.  This will give you official PSD where the area under the curve is 1.0.  I think there is also an option to do `cumulative=True`.  You can also just get the histogram from numpy using `np.hist` then you have more direct access to the data so you can fit curves for instance, then compare the fitting parameters of different situations, instead of doing max/min/mean as I think you were suggesting.

**Comment by fsamaraie** [comment]
Many thanks sir.

**Comment by AhmedTaiea** [comment]
Hello, I would like to calculate the Pore Size Distribution (PSD) for a 3D geometry. The pores in the 3D structure are represented as spheres within a TPMS (Triply Periodic Minimal Surface) structure. Additionally, I'm interested in extracting the network and connectivity of the pores. Since I'm not familiar with Python, any assistance with these tasks would be greatly appreciated. Thank you

**Reply by jgostick** [reply]
Have you have a scan through the examples [here](https://porespy.org/_examples/index.html)?  There are some tutorials as well as fairly detailed explanations for how each function works.  The functions are all categorized to help you find what you're looking for.

**Reply by AhmedTaiea** [reply]
Thanks for your reply Prof. I already tried reading and doing the examples, but the PSD I obtained using the pore_size_distribution function is in 2D and my structure is #D. Should I proceed with network extraction and extract the pores as spheres to calculate the PSD? Thanks.
(maybe my question looks stupid but that is because first time using Python and coding)

**Reply by jgostick** [reply]
Um, why would the distribution be in 3D?  Isn't a plot of pore size vs frequency a 2D graph or bar chart?  I must be missing something.  [Here](https://porespy.org/examples/filters/tutorials/local_thickness.html) is relevant example.

**Reply by AhmedTaiea** [reply]
I apologize for the misunderstanding. In the example you provided, a geometry was generated in the x and y dimensions using the code im = np.zeros([300, 300]). However, I would like to generate a similar geometry, but in three dimensions, including the x, y, and z dimensions. and after that calculate the PSD. Thank you for your understanding.

**Reply by AhmedTaiea** [reply]
Here you can see that my 3D cubic structure consists of solid spheres. I'm interested in obtaining the pore diameter distribution between these spheres. Many thanks for your help
![3d solid spheres](https://github.com/PMEAL/OpenPNM/assets/86796551/5a4ca2e4-063c-43bb-8936-9edb34864ec6)

**Comment by AhmedTaiea** [comment]
to have something like this in the end 
![materials-13-01779-g004](https://github.com/PMEAL/OpenPNM/assets/86796551/8785c249-4200-44df-a603-a46ead6a7ffa)
